# GSB 5544 — PA 1.2 Intro to Pandas — SOLUTION

**First save a copy of this notebook in your Google drive.** Read this notebook from top to bottom and fill in the code as you go. Work together with your partner and discuss with other students in the class. Try to resolve any errors on your own first, but don't get stuck; ask for help!

For many of these questions, there are several ways to do it in Pandas. You only need to choose one that works. But you might want to experiment with multiple ways of doing the same thing to get practice working with Pandas.





## Titanic Data

 The Titanic data set (https://raw.githubusercontent.com/kevindavisross/data301/main/data/titanic.csv) contains information about 2206 passengers and crew members that were aboard the RMS Titanic when it sank on April 15, 1912.

1\. Read in the Titanic data set. What seems to be the observational unit?

In [1]:
# YOUR CODE HERE - ADD AS MANY CELLS AS YOU WANT
import pandas as pd

df_titanic = pd.read_csv("https://raw.githubusercontent.com/kevindavisross/data301/main/data/titanic.csv")
print(df_titanic.shape)
df_titanic.head()

(2207, 11)


,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
0,"Abbing, Mr. Anthony",male,42.0,3rd,S,United States,5547.0,7.11,0,3.0,NaN
1,"Abbott, Mr. Eugene Joseph",male,13.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN
2,"Abbott, Mr. Rossmore Edward",male,16.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN
3,"Abbott, Mrs. Rhoda Mary 'Rosa'",female,39.0,3rd,S,England,2673.0,20.05,1,3.0,NaN
4,"Abelseth, Miss. Karen Marie",female,16.0,3rd,S,Norway,348125.0,7.13,1,3.0,NaN


**Answer:** Each row is one *person* aboard the Titanic (passenger or crew member) — 2,207 rows, 11 columns. The observational unit is a person.

2\. What column seems to be an appropriate index for this data frame? Can you think of any problems with using this column as the index?

**Answer:** `name` is the natural choice — it's the only column that identifies a person. Problem: names are **not unique** (e.g. there are three people named "Kelly, Mr. James"), so a name-indexed lookup can return several rows. `ticketno` doesn't work either: crew have no ticket, and families share tickets.

3\. Regardless of your answer to the previous question, write code to make "name" the index of the data frame.

In [2]:
# YOUR CODE HERE
df_titanic = df_titanic.set_index("name")
df_titanic.head()

,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
name,,,,,,,,,,
"Abbing, Mr. Anthony",male,42.0,3rd,S,United States,5547.0,7.11,0,3.0,NaN
"Abbott, Mr. Eugene Joseph",male,13.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN
"Abbott, Mr. Rossmore Edward",male,16.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN
"Abbott, Mrs. Rhoda Mary 'Rosa'",female,39.0,3rd,S,England,2673.0,20.05,1,3.0,NaN
"Abelseth, Miss. Karen Marie",female,16.0,3rd,S,Norway,348125.0,7.13,1,3.0,NaN


 4\. Write code that returns the observations for  "Kelly, Mr. James". What do you notice? (You can also try "Green, Mr. George"). Note: you might try this both with "name" as the index and with it not as the index.

In [3]:
# YOUR CODE HERE
# With "name" as the index -> .loc by label
df_titanic.loc["Kelly, Mr. James"]

,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
name,,,,,,,,,,
"Kelly, Mr. James",male,19.0,3rd,S,Scotland,363592.0,8.0100,0,3.0,NaN
"Kelly, Mr. James",male,44.0,3rd,Q,Ireland,330911.0,7.1607,0,3.0,NaN
"Kelly, Mr. James",male,44.0,engineering crew,S,England,NaN,NaN,0,NaN,engineering crew


**Answer:** Three different people share the name "Kelly, Mr. James" (ages 19 and 44, from Scotland, Ireland, and England — one is engineering crew). `.loc` returns a `DataFrame` with 3 rows rather than a single row, which is why `name` is a shaky index.

In [4]:
# Without "name" as the index -> filter on the column
df_titanic.reset_index()[df_titanic.reset_index()["name"] == "Kelly, Mr. James"]

,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
651,"Kelly, Mr. James",male,19.0,3rd,S,Scotland,363592.0,8.0100,0,3.0,NaN
652,"Kelly, Mr. James",male,44.0,3rd,Q,Ireland,330911.0,7.1607,0,3.0,NaN
653,"Kelly, Mr. James",male,44.0,engineering crew,S,England,NaN,NaN,0,NaN,engineering crew


5\. Write code to extract a single `DataFrame` containing the observations three members of the Widener family:
  - Widener, Mr. George Dunton
  - Widener, Mr. Harry Elkins
  - Widener, Mrs. Eleanor

What became of them? Using a search engine, what else can you learn about them? (Note: you could just provide a list of these names. But you might also want to experiment with functions for working with strings, e.g., [`str.contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html) to find all instances of "Widener"; in this case, it might be helpful to have name as a regular column and not the index.)

In [5]:
# YOUR CODE HERE
# Option 1: a list of labels
df_titanic.loc[["Widener, Mr. George Dunton", "Widener, Mr. Harry Elkins", "Widener, Mrs. Eleanor"]]

,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
name,,,,,,,,,,
"Widener, Mr. George Dunton",male,50.0,1st,S,United States,113503.0,211.1,0,1.0,NaN
"Widener, Mr. Harry Elkins",male,27.0,1st,S,United States,113503.0,211.1,0,1.0,NaN
"Widener, Mrs. Eleanor",female,50.0,1st,S,United States,113503.0,211.1,1,1.0,NaN


In [6]:
# Option 2: string matching on a regular column
df_titanic = df_titanic.reset_index()          # make "name" a column again
df_titanic[df_titanic["name"].str.contains("Widener")]

,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
1283,"Widener, Mr. George Dunton",male,50.0,1st,S,United States,113503.0,211.1,0,1.0,NaN
1284,"Widener, Mr. Harry Elkins",male,27.0,1st,S,United States,113503.0,211.1,0,1.0,NaN
1285,"Widener, Mrs. Eleanor",female,50.0,1st,S,United States,113503.0,211.1,1,1.0,NaN


**Answer:** George (50) and his son Harry (27) died; Eleanor (50) survived. George Widener was a Philadelphia streetcar magnate and one of the wealthiest passengers; Harry was a book collector — Harvard's Widener Library was built in his memory by Eleanor.

6\. Look at a few rows in the full data set. Try to identify what each variable measures. Also, try to identify each variable as either quantitative, categorical, or other. (In general, there should be a code book with information about all the variables. But try to infer what you can from context.)

In [7]:
# YOUR CODE HERE
df_titanic.sample(5, random_state=1)

,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew
1784,"Light, Mr. Christopher William",male,20.0,engineering crew,S,England,NaN,NaN,0,NaN,engineering crew
1578,"Fellowes, Mr. Alfred James",male,29.0,victualling crew,B,England,NaN,NaN,0,NaN,victualling crew
1231,"Vanderplancke, Mr. Julius",male,31.0,3rd,S,Belgium,345763.0,18.0000,0,3.0,NaN
415,"Francatelli, Miss. Laura Mabel",female,32.0,1st,C,England,17485.0,56.1807,1,1.0,NaN
201,"Carlsson, Mr. Frans Olof",male,33.0,1st,S,United States,695.0,5.0000,0,1.0,NaN


**Answer:**

| Variable | Measures | Type |
|---|---|---|
| `name` | passenger/crew name | identifier (other) |
| `gender` | male / female | categorical |
| `age` | age in years | quantitative |
| `class` | 1st/2nd/3rd or crew department | categorical |
| `embarked` | port of boarding (S, C, Q, B) | categorical |
| `country` | home country | categorical |
| `ticketno` | ticket number | identifier (other) — a number, but arithmetic is meaningless |
| `fare` | ticket price (£) | quantitative |
| `survived` | 0 = died, 1 = survived | categorical (stored as int) |
| `pclass` | passenger class 1/2/3 | categorical (stored as float) |
| `crew` | crew department, NaN for passengers | categorical |

7\. The variable `pclass` measures the class for passengers. Why are some of the values in this column `NaN`? Should this variable be categorical or quantitative? What type does Pandas this it is?

In [8]:
# YOUR CODE HERE
print(df_titanic["pclass"].dtype)
print(df_titanic["pclass"].value_counts(dropna=False))
df_titanic.loc[df_titanic["pclass"].isna(), "class"].value_counts()

float64
pclass
NaN    890
3.0    709
1.0    324
2.0    284
Name: count, dtype: int64


class
victualling crew    431
engineering crew    324
restaurant staff     69
deck crew            66
Name: count, dtype: int64

**Answer:** The 890 `NaN` values are all **crew members** — they don't have a passenger class. `pclass` should be **categorical** (1st/2nd/3rd are labels; an "average class" of 2.1 means nothing). Pandas thinks it is quantitative (`float64`) because it's stored as numbers, and the `NaN`s force float rather than int.

8\. Write code to cast `pclass` to the correct type. Repeat for any other variables that do not have the correct type.

In [9]:
# YOUR CODE HERE
df_titanic["pclass"]   = df_titanic["pclass"].astype("category")
df_titanic["survived"] = df_titanic["survived"].astype("category")
for col in ["gender", "class", "embarked", "country", "crew"]:
    df_titanic[col] = df_titanic[col].astype("category")
df_titanic["ticketno"] = df_titanic["ticketno"].astype(str)    # an ID, not a number

df_titanic.dtypes

name          object
gender      category
age          float64
class       category
embarked    category
country     category
ticketno      object
fare         float64
survived    category
pclass      category
crew        category
dtype: object

## Transforming Categorical Variables

A categorical variable can be transformed by mapping its levels to new levels. For example, we may only be interested in whether a person on the titanic was a passenger or a crew member. The variable `class` is too detailed for these purposes. We can create a new variable, `type`, that is derived from the existing variable `class`. Observations with a `class` of "1st", "2nd", or "3rd" get a value of "passenger", while observations with a `class` of "deck crew", "engineering crew", or "deck crew" get a value of "crew".

In [10]:
df_titanic["type"] = df_titanic["class"].map({
    "1st": "passenger",
    "2nd": "passenger",
    "3rd": "passenger",
    "victualling crew": "crew",
    "engineering crew": "crew",
    "deck crew": "crew"
})

df_titanic

,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew,type
0,"Abbing, Mr. Anthony",male,42.0,3rd,S,United States,5547.0,7.11,0,3.0,NaN,passenger
1,"Abbott, Mr. Eugene Joseph",male,13.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN,passenger
2,"Abbott, Mr. Rossmore Edward",male,16.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN,passenger
3,"Abbott, Mrs. Rhoda Mary 'Rosa'",female,39.0,3rd,S,England,2673.0,20.05,1,3.0,NaN,passenger
4,"Abelseth, Miss. Karen Marie",female,16.0,3rd,S,Norway,348125.0,7.13,1,3.0,NaN,passenger
...,...,...,...,...,...,...,...,...,...,...,...,...
2202,"Wynn, Mr. Walter",male,41.0,deck crew,B,England,nan,NaN,1,NaN,deck crew,crew
2203,"Yearsley, Mr. Harry",male,40.0,victualling crew,S,England,nan,NaN,1,NaN,victualling crew,crew
2204,"Young, Mr. Francis James",male,32.0,engineering crew,S,England,nan,NaN,0,NaN,engineering crew,crew
2205,"Zanetti, Sig. Minio",male,20.0,restaurant staff,S,England,nan,NaN,0,NaN,restaurant staff,NaN


Upon closer inspection of this `DataFrame`, we see that we accidentally left out the level "restaurant staff" in the input to `.map()`. Any levels that are unspecified will be mapped to the missing value `NaN`.

This suggests a more concise way to define the new variable `type`. We can specify only the levels for passengers in the mapping and then fill in the missing values afterwards.

In [11]:
df_titanic["type"] = df_titanic["class"].map({
    "1st": "passenger",
    "2nd": "passenger",
    "3rd": "passenger"
})

# Replace all missing values by "crew"
df_titanic["type"] = df_titanic["type"].fillna("crew")

df_titanic

,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew,type
0,"Abbing, Mr. Anthony",male,42.0,3rd,S,United States,5547.0,7.11,0,3.0,NaN,passenger
1,"Abbott, Mr. Eugene Joseph",male,13.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN,passenger
2,"Abbott, Mr. Rossmore Edward",male,16.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN,passenger
3,"Abbott, Mrs. Rhoda Mary 'Rosa'",female,39.0,3rd,S,England,2673.0,20.05,1,3.0,NaN,passenger
4,"Abelseth, Miss. Karen Marie",female,16.0,3rd,S,Norway,348125.0,7.13,1,3.0,NaN,passenger
...,...,...,...,...,...,...,...,...,...,...,...,...
2202,"Wynn, Mr. Walter",male,41.0,deck crew,B,England,nan,NaN,1,NaN,deck crew,crew
2203,"Yearsley, Mr. Harry",male,40.0,victualling crew,S,England,nan,NaN,1,NaN,victualling crew,crew
2204,"Young, Mr. Francis James",male,32.0,engineering crew,S,England,nan,NaN,0,NaN,engineering crew,crew
2205,"Zanetti, Sig. Minio",male,20.0,restaurant staff,S,England,nan,NaN,0,NaN,restaurant staff,crew


For more complex mappings, the `.map()` method also accepts a function. So the above is equivalent to

In [12]:
def class_to_type(c):
  if c in ["1st", "2nd", "3rd"]:
    return "passenger"
  else:
    return "crew"

df_titanic["class"].map(class_to_type)

0       passenger
1       passenger
2       passenger
3       passenger
4       passenger
          ...    
2202         crew
2203         crew
2204         crew
2205         crew
2206         crew
Name: class, Length: 2207, dtype: object

## Converting Quantitative Variables to Categorical Variables

Sometimes it is necessary to categorize quantitative variables. We can use the `cut` function in `pandas` to categorize or bin a quantitative variable, by specifying the endpoints of the bins. We can also provide labels for the bins.

In [13]:
df_titanic["age_categories"] = pd.cut(x = df_titanic["age"],
                                      bins = [0, 18, 30, 45, 65, 130],
                                      labels = ["<18", "18-30", "30-45", "45-65", "65+"])

df_titanic

,name,gender,age,class,embarked,country,ticketno,fare,survived,pclass,crew,type,age_categories
0,"Abbing, Mr. Anthony",male,42.0,3rd,S,United States,5547.0,7.11,0,3.0,NaN,passenger,30-45
1,"Abbott, Mr. Eugene Joseph",male,13.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN,passenger,<18
2,"Abbott, Mr. Rossmore Edward",male,16.0,3rd,S,United States,2673.0,20.05,0,3.0,NaN,passenger,<18
3,"Abbott, Mrs. Rhoda Mary 'Rosa'",female,39.0,3rd,S,England,2673.0,20.05,1,3.0,NaN,passenger,30-45
4,"Abelseth, Miss. Karen Marie",female,16.0,3rd,S,Norway,348125.0,7.13,1,3.0,NaN,passenger,<18
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2202,"Wynn, Mr. Walter",male,41.0,deck crew,B,England,nan,NaN,1,NaN,deck crew,crew,30-45
2203,"Yearsley, Mr. Harry",male,40.0,victualling crew,S,England,nan,NaN,1,NaN,victualling crew,crew,30-45
2204,"Young, Mr. Francis James",male,32.0,engineering crew,S,England,nan,NaN,0,NaN,engineering crew,crew,30-45
2205,"Zanetti, Sig. Minio",male,20.0,restaurant staff,S,England,nan,NaN,0,NaN,restaurant staff,crew,18-30


Now we can treat the new variable as we would any other categorical variable.

In [14]:
df_titanic["age_categories"]

0       30-45
1         <18
2         <18
3       30-45
4         <18
        ...  
2202    30-45
2203    30-45
2204    30-45
2205    18-30
2206    18-30
Name: age_categories, Length: 2207, dtype: category
Categories (5, object): ['<18' < '18-30' < '30-45' < '45-65' < '65+']

Arguments to the `cut` method can be specified to determine the treatment of the bin endpoints, but we won't go into the details here. There is also a `qcut` method which creates bins based on quantiles (e.g., the endpoints are the 25th, 50th, and 75th percentiles (along with the min and max).)

## Back to the Exercises about the Titanic data...

10\. Create a new country column that contains only the values United States, United Kingdom, or Other. (At the time of the Titanic, the countries in the United Kingdom were England, Scotland, Wales, and Ireland, including what is now Northern Ireland.) Don't overwrite the original country column.

In [15]:
# YOUR CODE HERE
uk = ["England", "Scotland", "Wales", "Ireland", "Northern Ireland"]

df_titanic["country_group"] = df_titanic["country"].astype(str).map(
    {"United States": "United States", **{c: "United Kingdom" for c in uk}}
).fillna("Other").astype("category")

df_titanic["country_group"].value_counts()

country_group
United Kingdom    1339
Other              604
United States      264
Name: count, dtype: int64

11\. Create a new column that categorizes the fare variable into bins of length 5 from 0 to 25, bins of length 25 from 25 to 100, and 100+.

In [16]:
# YOUR CODE HERE
df_titanic["fare_bin"] = pd.cut(df_titanic["fare"],
                                bins=[0, 5, 10, 15, 20, 25, 50, 75, 100, 1000],
                                labels=["0-5", "5-10", "10-15", "15-20", "20-25",
                                        "25-50", "50-75", "75-100", "100+"])

df_titanic["fare_bin"].value_counts().sort_index()

fare_bin
0-5         3
5-10      473
10-15     186
15-20      73
20-25      69
25-50     247
50-75      96
75-100     61
100+       83
Name: count, dtype: int64

## Working with Gemini

Colab notebooks have [Gemini](), Google's AI, built in. Let's try it.

1\. Load either the Titanic data frame. After you load the data frame, there should be a button for "View recommended plot". Turn that button on and see what you get. Select one of the suggested plots that summarizes a single categorical variable. (Careful: make sure it's a single categorical variable. Some of the bar plots might be summarizing multiple variables.) What are features that you like about the plot? Are there any features you don't like?

2\. Now click on the "generate" with AI link, or the Gemini button. Write a prompt to reproduce some of your analysis from earlier in this notebook. You can copy and paste one of the questions, or you can try writing your own prompt. Run the code and examine the results. How successful was it?